File: app/src/main/java/org/cosmic/ide/activity/ProjectActivity.java
package org.cosmic.ide.activity;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import androidx.annotation.WorkerThread;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.github.pedrovgs.lynx.LynxActivity;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.materialswitch.MaterialSwitch;

import org.cosmic.ide.R;
import org.cosmic.ide.activity.adapter.ProjectAdapter;
import org.cosmic.ide.common.util.CoroutineUtil;
import org.cosmic.ide.databinding.ActivityProjectBinding;
import org.cosmic.ide.databinding.DialogNewProjectBinding;
import org.cosmic.ide.git.model.Author;
import org.cosmic.ide.git.usecases.UseCasesKt;
import org.cosmic.ide.project.JavaProject;
import org.cosmic.ide.project.KotlinProject;
import org.cosmic.ide.project.Project;
import org.cosmic.ide.util.AndroidUtilities;
import org.cosmic.ide.util.Constants;
import org.cosmic.ide.util.UiUtilsKt;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

public class ProjectActivity extends BaseActivity<ActivityProjectBinding>
        implements ProjectAdapter.OnProjectEventListener {

    public interface OnProjectCreatedListener {
        void onProjectCreated(Project project);
    }

    private ProjectAdapter projectAdapter;

    private AlertDialog createNewProjectDialog;
    private DialogNewProjectBinding projectBinding;

    private OnProjectCreatedListener mListener;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(ActivityProjectBinding.inflate(getLayoutInflater()));
        setSupportActionBar(binding.toolbar);

        buildCreateNewProjectDialog();

        projectAdapter = new ProjectAdapter();
        binding.projectRecycler.setAdapter(projectAdapter);
        binding.projectRecycler.setLayoutManager(new LinearLayoutManager(this));
        binding.projectRecycler.addItemDecoration(new DividerItemDecoration(this, DividerItemDecoration.VERTICAL));
        projectAdapter.setOnProjectEventListener(this);
        mListener = this::onProjectClicked;

        UiUtilsKt.addSystemWindowInsetToMargin(binding.fab, false, false, false, true);
        UiUtilsKt.addSystemWindowInsetToPadding(binding.appBar, false, true, false, false);
        UiUtilsKt.addSystemWindowInsetToPadding(binding.projectRecycler, false, false, false, true);

        binding.refreshLayout.setOnRefreshListener(
                () -> {
                    loadProjects();
                    binding.refreshLayout.setRefreshing(false);
                });
        binding.fab.setOnClickListener(v -> showCreateNewProjectDialog());
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.projects_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        final var id = item.getItemId();
        if (id == R.id.action_settings) {
            startActivity(new Intent(this, SettingsActivity.class));
        } else if (id == R.id.action_logcat) {
            startActivity(LynxActivity.getIntent(this));
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onResume() {
        super.onResume();
        loadProjects();
    }

    @Override
    public int getLayoutRes() {
        return R.layout.activity_project;
    }

    private void buildCreateNewProjectDialog() {
        var builder =
                new MaterialAlertDialogBuilder(
                                this, AndroidUtilities.getDialogFullWidthButtonsThemeOverlay())
                        .setTitle(getString(R.string.create_project));
        projectBinding = DialogNewProjectBinding.inflate(LayoutInflater.from(builder.getContext()));
        builder
                .setView(projectBinding.getRoot())
                .setPositiveButton(getString(R.string.create), null)
                .setNegativeButton(getString(android.R.string.cancel), null);
        createNewProjectDialog = builder.create();
    }

    @WorkerThread
    private void showCreateNewProjectDialog() {
        if (!createNewProjectDialog.isShowing()) {
            createNewProjectDialog.show();
            Button createBtn = createNewProjectDialog.findViewById(android.R.id.button1);
            createBtn.setOnClickListener(
                    v -> {
                        var projectName = projectBinding.text1.getText().toString().trim().replace("..", "");
                        if (projectName.isEmpty()) {
                            return;
                        }
                        boolean useKotlinTemplate = projectBinding.useKotlinTemplate.isChecked();

                        try {
                            var project =
                                    useKotlinTemplate
                                            ? KotlinProject.newProject(projectName)
                                            : JavaProject.newProject(projectName);
                            if (projectBinding.useGit.isChecked()) {
                                final var author = new Author(getSettings().getGitUserName(), getSettings().getGitUserEmail());
                                UseCasesKt.createGitRepoWith(project.getProjectDirPath(), author, "Initial Commit");
                            }
                            if (mListener != null) {
                                runOnUiThread(
                                        () -> {
                                            if (createNewProjectDialog.isShowing())
                                                createNewProjectDialog.dismiss();
                                            mListener.onProjectCreated(project);
                                        });
                            }
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                        loadProjects();
                    });
            projectBinding.text1.setText("");
        }
    }

    @WorkerThread
    private void showDeleteProjectDialog(Project project) {
        AndroidUtilities.showSimpleAlert(
                this,
                getString(R.string.dialog_delete),
                getString(R.string.dialog_confirm_delete, project.getProjectName()),
                getString(android.R.string.yes),
                getString(android.R.string.no),
                ((dialog, which) -> {
                    if (which == DialogInterface.BUTTON_POSITIVE) {
                        project.delete();
                        runOnUiThread(this::loadProjects);
                    }
                }));
    }

    @Override
    public void onProjectClicked(Project project) {
        var projectPath = project.getProjectDirPath();
        var intent = new Intent(this, MainActivity.class);
        intent.putExtra(Constants.PROJECT_PATH, projectPath);
        startActivity(intent);
    }

    @Override
    public boolean onProjectLongClicked(Project project) {
        showDeleteProjectDialog(project);
        return true;
    }

    @WorkerThread
    private void loadProjects() {
        CoroutineUtil.inParallel(
                () -> {
                    var projectDir = new File(JavaProject.getRootDirPath());
                    var directories = projectDir.listFiles(File::isDirectory);
                    var projects = new ArrayList<Project>();
                    if (directories != null) {
                        Arrays.sort(directories, Comparator.comparingLong(File::lastModified));
                        for (var directory : directories) {
                            var project = new File(directory, "src");
                            if (project.exists()) {
                                var javaProject =
                                        new JavaProject(new File(directory.getAbsolutePath()));
                                projects.add(javaProject);
                            }
                        }
                    }
                    runOnUiThread(
                            () -> {
                                projectAdapter.submitList(projects);
                                toggleNullProject(projects);
                            });
                });
    }

    private void toggleNullProject(List<Project> projects) {
        if (projects.size() == 0) {
            binding.projectRecycler.setVisibility(View.GONE);
            binding.emptyContainer.setVisibility(View.VISIBLE);
        } else {
            binding.projectRecycler.setVisibility(View.VISIBLE);
            binding.emptyContainer.setVisibility(View.GONE);
        }
    }
}

File: app/src/main/java/org/cosmic/ide/activity/ProjectActivity.kt
package org.cosmic.ide.activity

import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Button
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.WorkerThread
import androidx.appcompat.app.AlertDialog
import androidx.documentfile.provider.DocumentFile
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.github.pedrovgs.lynx.LynxActivity
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import org.cosmic.ide.R
import org.cosmic.ide.activity.adapter.ProjectAdapter
import org.cosmic.ide.activity.adapter.ProjectAdapter.OnProjectEventListener
import org.cosmic.ide.common.util.CoroutineUtil.inParallel
import org.cosmic.ide.common.util.FileUtil
import org.cosmic.ide.common.util.unzip
import org.cosmic.ide.databinding.ActivityProjectBinding
import org.cosmic.ide.databinding.DialogNewProjectBinding
import org.cosmic.ide.git.model.Author
import org.cosmic.ide.git.usecases.createGitRepoWith
import org.cosmic.ide.project.JavaProject
import org.cosmic.ide.project.KotlinProject
import org.cosmic.ide.project.Project
import org.cosmic.ide.util.AndroidUtilities
import org.cosmic.ide.util.Constants
import org.cosmic.ide.util.addSystemWindowInsetToMargin
import org.cosmic.ide.util.addSystemWindowInsetToPadding
import java.io.File
import java.io.IOException
import java.util.Arrays
import java.util.Comparator

class ProjectActivity : BaseActivity(), OnProjectEventListener {
    private val REQUEST_CODE_SELECT_PROJECT = 0
    private val projectAdapter = ProjectAdapter()
    private val projectBinding by lazy {
        DialogNewProjectBinding.inflate(layoutInflater)
    }
    private val createNewProjectDialog: AlertDialog by lazy {
        val dialog = MaterialAlertDialogBuilder(
            this, AndroidUtilities.dialogFullWidthButtonsThemeOverlay
        ).apply {
            setTitle(getString(R.string.create_project))
            setView(projectBinding.root)
            setPositiveButton(getString(R.string.create), null)
            setNegativeButton(getString(android.R.string.cancel), null)
        }
        dialog.create()
    }

    private val mListener = object : OnProjectCreatedListener {
        override fun onProjectCreated(project: Project) {
            onProjectClicked(project)
        }
    }

    private lateinit var binding: ActivityProjectBinding

    interface OnProjectCreatedListener {
        fun onProjectCreated(project: Project)
    }
    private val importListener = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
                    if (result.resultCode == RESULT_OK) {
                        result.data?.data?.let { uri ->
                            val file = DocumentFile.fromSingleUri(this, uri)!!
                            val name = file.name?.substringBeforeLast(".")
                            val path = FileUtil.getProjectsDir() + name
                            val projectDir = File(path)
                            if (projectDir.exists()) {
                                MaterialAlertDialogBuilder(this)
                                    .setTitle("Project already exists")
                                    .setMessage("Do you want to overwrite it?")
                                    .setPositiveButton("Yes") { _, _ ->
                                        projectDir.deleteRecursively()
                                        contentResolver.openInputStream(uri)!!.use {
                                            it.unzip(File(FileUtil.getProjectsDir()))
                                            loadProjects()
                                        }
                                    }
                                    .setNegativeButton("No") { _, _ -> }
                                    .show()
                            } else {
                                contentResolver.openInputStream(uri)!!.use {
                                    it.unzip(projectDir)
                                    loadProjects()
                                }
                            }
                        }
                    }
                }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityProjectBinding.inflate(layoutInflater).also {
            setContentView(it.root)
            setSupportActionBar(it.toolbar)
        }
        binding.projectRecycler.apply {
            adapter = projectAdapter
            layoutManager = LinearLayoutManager(this@ProjectActivity)
            addItemDecoration(
                DividerItemDecoration(
                    this@ProjectActivity,
                    DividerItemDecoration.VERTICAL
                )
            )
            addSystemWindowInsetToPadding(bottom = true)
        }
        binding.apply {
            fab.addSystemWindowInsetToMargin(bottom = true)
            appBar.addSystemWindowInsetToPadding(top = true)
            refreshLayout.setOnRefreshListener {
                loadProjects()
                refreshLayout.isRefreshing = false
            }
            fab.setOnClickListener {
                showCreateNewProjectDialog()
            }
        }
        projectAdapter.onProjectEventListener = this@ProjectActivity
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.projects_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_settings -> {
                startActivity(Intent(this, SettingsActivity::class.java))
            }

            R.id.action_logcat -> {
                startActivity(LynxActivity.getIntent(this))
            }

            R.id.import_project -> {
                // Create an intent for the user to select a directory
                val intent = Intent(Intent.ACTION_OPEN_DOCUMENT).apply {
                    type = "application/zip"
                }
                importListener.launch(intent)
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onResume() {
        super.onResume()
        loadProjects()
    }

    @WorkerThread
    private fun showCreateNewProjectDialog() {
        if (!createNewProjectDialog.isShowing) {
            createNewProjectDialog.show()
            val createBtn: Button = createNewProjectDialog.findViewById(android.R.id.button1)!!
            createBtn.setOnClickListener {
                val projectName =
                    projectBinding.text1.text.toString().trim().replace("..", "")
                if (projectName.isEmpty()) {
                    return@setOnClickListener
                }
                val useKotlinTemplate =
                    projectBinding.useKotlinTemplate.isChecked
                try {
                    val project = if (useKotlinTemplate) {
                        KotlinProject.newProject(projectName)
                    } else {
                        JavaProject.newProject(projectName)
                    }
                    if (projectBinding.useGit.isChecked) {
                        val author = Author(settings.gitUserName, settings.gitUserEmail)
                        project.projectDirPath.createGitRepoWith(
                            author,
                            "Initial Commit"
                        )
                    }
                    runOnUiThread {
                        if (createNewProjectDialog.isShowing) {
                            createNewProjectDialog.dismiss()
                        }
                        mListener.onProjectCreated(project)
                    }
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
            projectBinding.text1.setText("")
        }
    }

    @WorkerThread
    private fun showDeleteProjectDialog(project: Project) {
        AndroidUtilities.showSimpleAlert(
            this,
            getString(R.string.dialog_delete),
            getString(R.string.dialog_confirm_delete, project.projectName),
            getString(android.R.string.ok),
            getString(android.R.string.cancel)
        ) { _, which ->
            if (which == DialogInterface.BUTTON_POSITIVE) {
                project.delete()
                loadProjects()
            }
        }
    }

    override fun onProjectClicked(project: Project) {
        val projectPath = project.projectDirPath
        val intent = Intent(this, MainActivity::class.java)
        intent.putExtra(Constants.PROJECT_PATH, projectPath)
        startActivity(intent)
    }

    override fun onProjectLongClicked(project: Project): Boolean {
        showDeleteProjectDialog(project)
        return true
    }

    @WorkerThread
    private fun loadProjects() {
        inParallel {
            val projectDir = File(JavaProject.getRootDirPath())
            val directories =
                projectDir.listFiles { file -> file.isDirectory }
            val projects = mutableListOf<Project>()
            if (directories != null) {
                Arrays.sort(directories, Comparator.comparingLong(File::lastModified).reversed())
                for (directory in directories) {
                    projects.add(JavaProject(directory))
                }
            }
            runOnUiThread {
                projectAdapter.submitList(projects)
                toggleNullProject(projects)
            }
        }
    }

    private fun toggleNullProject(projects: List<Project>) {
        if (projects.isEmpty()) {
            binding.projectRecycler.visibility = View.GONE
            binding.emptyContainer.visibility = View.VISIBLE
        } else {
            binding.projectRecycler.visibility = View.VISIBLE
            binding.emptyContainer.visibility = View.GONE
        }
    }
}
